### ISSUE TITLE
ACTool task fails build when xcrun actool cannot find tool

### ISSUE THREAD

**Issue by AArnott** [xamarin/xamarin-macios#21762]
### Apple platform

iOS

### Framework version

net8.0-*, net9.0-*

### Affected platform version

.NET 9.0.100

### Description

actool cannot be found, but only when in the context of `dotnet build`:

> /usr/local/share/dotnet/packs/Microsoft.iOS.Sdk.net9.0_18.1/18.1.9163/tools/msbuild/iOS/Xamarin.Shared.targets(907,3): /usr/bin/xcrun exited with code 72:
xcrun: error: sh -c '/Applications/Xcode_16.app/Contents/Developer/usr/bin/xcodebuild -sdk /Applications/Xcode_16.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -find actool 2> /dev/null' failed with exit code 16384: (null) (errno=No such file or directory)
xcrun: error: unable to find utility "actool", not a developer tool or in PATH [/Volumes/Expansion/git/nerdbank.cryptocurrencies/src/Nerdbank.Zcash.App/Nerdbank.Zcash.App.iOS/Nerdbank.Zcash.App.iOS.csproj]

If I take exactly the command line from the error message and execute it directly:

    sh -c '/Applications/Xcode_16.app/Contents/Developer/usr/bin/xcodebuild -sdk /Applications/Xcode_16.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -find actool 2> /dev/null'

I get the path to the actool printed in stdout.

I suspect then that the problem is environment variables set by the ACTool msbuild task, but I can't tell because that information isn't captured by the binlog.

### Steps to Reproduce

This is a closed source application. 

### Did you find any workaround?

No.

### Build logs

binlog: [msbuild.zip](https://github.com/user-attachments/files/18029100/msbuild.zip)

**Comment by AArnott** [comment]
This happens exactly when I switch to Xcode 16.1 (which .NET 9 targeting requires). If I switch back to Xcode 16.0, it works just fine.

**Comment by rolfbjarne** [comment]
This looks like a variation of https://github.com/xamarin/xamarin-macios/issues/3931.

Can you try executing this script, to see if that shows the same error?

https://gist.github.com/rolfbjarne/0cfcd18e187102ad48402aefe2d0f2c4

> why does the .NET 9 SDK create a .NET 6 targeted maui project anyway??

That sounds like a different bug (I don't see this behavior). I would suggest filing this in the MAUI repo.

**Comment by AArnott** [comment]
@rolfbjarne That script appears to have worked when targeting Xcode 16.1, both with and without the leading `export` statements.

**Comment by rolfbjarne** [comment]
@AArnott and building your project with `dotnet build ...` from the command line still fails?

**Comment by AArnott** [comment]
@rolfbjarne yes.

**Comment by rolfbjarne** [comment]
Can you add this to your csproj:

```xml
<Target Name="_PrintEnvironment" BeforeTargets="_CoreCompileImageAssets">
	<Exec Command="env -0 | sort -z | tr '\0' '\n' > /tmp/actoolenv.txt" />
</Target>
```

build your project (reproducing the error), and then attach the resulting `/tmp/actoolenv.txt` file here?

**Comment by AArnott** [comment]
Sure. Here you go, @rolfbjarne:
[actoolenv.txt](https://github.com/user-attachments/files/18215748/actoolenv.txt)

**Comment by rolfbjarne** [comment]
> XCODE_DEVELOPER_DIR_PATH=/Applications/Xcode_16.1.app/Contents/Developer

Something is configured to use `/Applications/Xcode_16.1.app` instead of `/Applications/Xcode_16.app`, and that confuses Xcode.

Do you have this file: `~/Library/Preferences/Xamarin/Settings.plist`? If so, does deleting it fix the problem?

**Comment by AArnott** [comment]
Thanks, @rolfbjarne 

> Do you have this file: ~/Library/Preferences/Xamarin/Settings.plist? If so, does deleting it fix the problem?

That file does not exist.

> Something is configured to use /Applications/Xcode_16.1.app instead of /Applications/Xcode_16.app, and that confuses Xcode.

Both of these are symbolic links. In the `/Applications` directory I have:

```
Xcode.app -> Xcode_16.app
Xcode_16.app -> Xcode_16.0.app
Xcode_16.0.app -> /Volumes/Expansion/Xcode_16.0.app/
Xcode_16.1.app -> /Volumes/Expansion/Xcode_16.1.app/
```

`xcode-select -p` emits:

```
/Applications/Xcode_16.1.app/Contents/Developer
```

It is important as part of the repro that we build with Xcode 16.1, so yes, we're using `/Applications/Xcode_16.1.app` instead of `/Applications/Xcode_16.app`, and that's by design.

**Comment by rolfbjarne** [comment]
> > Something is configured to use /Applications/Xcode_16.1.app instead of /Applications/Xcode_16.app, and that confuses Xcode.
> 
> Both of these are symbolic links. In the `/Applications` directory I have:

I've seen Xcode getting confused when dealing with symbolic links. Can you try using the real path instead as the Xcode path?

    sudo xcode-select --switch /Volumes/Expansion/Xcode_16.1.app/

and see if that works?

**Comment by AArnott** [comment]
@rolfbjarne Quite right. Using the final path instead of the symbolic link resolved the issue. 
That's too bad though, because there's something of a contract as to where these apps are found when it comes to GitHub Actions or Azure Pipelines' agents, and yet the mac mini I have to build on doesn't have storage enough for Xcode to be installed on the main volume, forcing me to unpack Xcode on an expansion drive. I guess I'll just have to hard-code the proprietary path into my git repo.

Thank you for your help!

**Comment by rolfbjarne** [comment]
Thanks for confirming, I'll keep this open to see if we can detect that the Xcode path is a symlink and show a warning.

**Comment by jamescrosswell** [comment]
Just ran into this with Xcode 26. In our github actions we had:
```yaml
    # See https://github.com/actions/runner-images/blob/main/images/macos/macos-15-Readme.md#xcode
    - name: Pin the Xcode Version
      if: runner.os == 'macOS'
      shell: bash
      run: sudo xcode-select --switch /Applications/Xcode_26.app
```

In theory, that should work since there's a symlink from `/Applications/Xcode_26.0.app` to `/Applications/Xcode_26.0.1.app` in the [macos-15 runner](https://github.com/actions/runner-images/blob/main/images/macos/macos-15-Readme.md#xcode). 

In practice, dotnet fails to find things like actool so the above had to be changed so that `xcode-select` references the specific path rather than the symlink:
```yaml
    # See https://github.com/actions/runner-images/blob/main/images/macos/macos-15-Readme.md#xcode
    - name: Pin the Xcode Version
      if: runner.os == 'macOS'
      shell: bash
      run: sudo xcode-select --switch /Applications/Xcode_26.0.1.app
```

I guess that's OK once you know... took me a few hours to discover this though and the solution/workaround is a bit more fragile as, if GitHub bump the version of Xcode that is installed to 26.0.2 or 26.1.0 in the future, all our builds will start breaking. The symlink could survive minor version bumps.

**Comment by Mertsch** [comment]
MAUI .NET 10: It seems to be happening again...

https://github.com/actions/runner-images/blob/51665f334774c42174771890df8acafcf55010b6/images/macos/macos-15-Readme.md?plain=1#L166
```
| Version        | Build    | Path                                           | Symlinks                                                       |
| -------------- | -------- | ---------------------------------------------- | -------------------------------------------------------------- |
| 26.1           | 17B54    | /Applications/Xcode_26.1_Release_Candidate.app | /Applications/Xcode_26.1.0.app<br>/Applications/Xcode_26.1.app |
```

with `26.1`
```
xcodebuild: error: SDK "/Applications/Xcode_26.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk" cannot be located.
xcrun: error: sh -c '/Applications/Xcode_26.1.app/Contents/Developer/usr/bin/xcodebuild -sdk /Applications/Xcode_26.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -find actool 2> /dev/null' failed with exit code 16384: (null) (errno=No such file or directory)
xcrun: error: unable to find utility "actool", not a developer tool or in PATH
```

with `26.1_Release_Candidate` it works.

**Comment by jamescrosswell** [comment]
@Mertsch did you [try this](https://github.com/dotnet/macios/issues/21762#issuecomment-3424033859)?

**Comment by Mertsch** [comment]
> [@Mertsch](https://github.com/Mertsch) did you [try this](https://github.com/dotnet/macios/issues/21762#issuecomment-3424033859)?

Yes, that's what I am saying, symlink `/Applications/Xcode_26.1.app` does not work,
`/Applications/Xcode_26.1_Release_Candidate.app` does.

**Comment by reinierkops** [comment]
When we're using the /Applications/Xcode_26.1_Release_Candidate.app path, we're getting this error in the pipeline when building a .net10-ios application

`<dict>
  	<key>com.apple.actool.compilation-results</key>
  	<dict>
  		<key>output-files</key>
  		<array>
  			<string>/Users/runner/work/1/s/Source/obj/Develop/net10.0-ios/ios-arm64/actool/bundle/appicon60x60@2x.png</string>
  			<string>/Users/runner/work/1/s/Source/obj/Develop/net10.0-ios/ios-arm64/actool/bundle/appicon76x76@2x~ipad.png</string>
  			<string>/Users/runner/work/1/s/Source/obj/Develop/net10.0-ios/ios-arm64/actool/partial-info.plist</string>
  		</array>
  	</dict>
  	<key>com.apple.actool.errors</key>
  	<array>
  		<dict>
  			<key>description</key>
  			<string>No simulator runtime version from ["22E238", "22F77", "22G86", "23A8464"] available to use with iphonesimulator SDK version 23B77</string>
  		</dict>
  	</array>
  </dict>`

We're builing it for ios-arm64, without the use of the simulator...

**Comment by rolfbjarne** [comment]
> We're builing it for ios-arm64, without the use of the simulator...

Yes, for some reason Apple's toolchain requires simulator bits, even when building for device (in this case it's the `actool`, which is used to compile asset catalogs, but it's not the only tool that always requires the simulator).

And it may work to do this first:

> $ xcodebuild -downloadPlatform iOS

**Comment by SteveBush** [comment]
Is there a working example of building an MAUI iOS application in Azure Pipelines that targets .NET 10? I have not found an XCode SDK version that works, including the RC.  GitHub actions work fine when use image macOS 26 and XCode 26.1.0.

**Comment by zippo227** [comment]
> Is there a working example of building an MAUI iOS application in Azure Pipelines that targets .NET 10? I have not found an XCode SDK version that works, including the RC. GitHub actions work fine when use image macOS 26 and XCode 26.1.0.

I added some details here.

https://github.com/dotnet/macios/issues/24314#issuecomment-3568317459
